### ISSUE TITLE
Builtin make on macOS often does not work with any -j besides -j1 (Permission denied)

### ISSUE THREAD

**Issue by certik** [LLNL/spack#26669]
### Steps to reproduce

This is something that I have noticed on unrelated project: when I use the default `/usr/bin/make` it has the following behavior:
```console
lfortran/integration_tests(cpp14) $ make -j   
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: *** [CMakeFiles/return_04.dir/all] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/sin_03.dir/all] Error 1
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: *** [CMakeFiles/sin_01.dir/all] Error 1
make[1]: *** [CMakeFiles/abs_03.dir/all] Error 1
[  0%] Building Fortran object CMakeFiles/class_02.dir/class_02.f90.o
[  0%] Linking Fortran executable class_02
[  0%] Built target class_02
make: *** [all] Error 2
lfortran/integration_tests(cpp14) $ make -j3
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: *** [CMakeFiles/sin_03.dir/all] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/sin_01.dir/all] Error 1
[  0%] Building Fortran object CMakeFiles/abs_03.dir/abs_03.f90.o
[  1%] Linking Fortran executable abs_03
[  1%] Built target abs_03
make: *** [all] Error 2
lfortran/integration_tests(cpp14) $ make -j2
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: /Library/Developer/CommandLineTools/usr/bin/make: Permission denied
make[1]: *** [CMakeFiles/sin_03.dir/all] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/sin_01.dir/all] Error 1
make: *** [all] Error 2
lfortran/integration_tests(cpp14) $ make -j1
[  0%] Building Fortran object CMakeFiles/sin_03.dir/sin_03.f90.o
[  0%] Linking Fortran executable sin_03
[  0%] Built target sin_03
[  0%] Building Fortran object CMakeFiles/sin_01.dir/sin_01.f90.o
[  0%] Linking Fortran executable sin_01
[  0%] Built target sin_01
[  1%] Built target abs_03
[  1%] Built target class_02
[  1%] Building Fortran object CMakeFiles/return_04.dir/return_04.f90.o
[  2%] Linking Fortran executable return_04
[  2%] Built target return_04
[  2%] Building Fortran object CMakeFiles/return_02.dir/return_02.f90.o
[  3%] Linking Fortran executable return_02
[  3%] Built target return_02
[  3%] Building Fortran object CMakeFiles/return_01.dir/return_01.f90.o
[  3%] Linking Fortran executable return_01
[  3%] Built target return_01
...
```

But if I use the `make` from Conda, then it works just fine in parallel.

I am reporting this as a standalone issue, because this seems to be happening on both of my two macs, the MacBook Pro 2019 (Intel) and 2020 (M1). As a consequence, this then triggers hard to debug errors, in particular:

* #25598
* https://github.com/spack/spack/issues/26667#issuecomment-940520983

They are all caused by the fact that Spack calls the make from macOS with -j8 (or whatever) and it fails with permission issues. However you can't tell that they are coming from make itself! Very hard to debug.

It's probably some permission issue, but what could it be? It would be nice to figure out, now when we have isolated the bug.

And a workaround is also super simple: just use `spack install -j1`.  Now, it gets tricky, because due to this Spack also does not bootstrap, but one can just see what packages it is trying to build, and then build them directly with `spack install -j1 zlib` and such. That way it seems one can build everything.

### Error message

N/A

### Information on your system

N/A

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
That's odd, I'm also using the builtin `make` on macOS 10.15.7 and it works fine in parallel. Maybe you installed the macOS command line tools (or XCode) without accepting the license agreement? You should be prompted to do this the first time you open XCode.

**Comment by certik** [comment]
I installed the command line tools by typing `git` and hitting enter in a terminal. A dialog pops up asking me to download the tools, then I accept the agreement and it installs the tools.

If you have any ideas how to dig at this issue deeper, please let me know.

**Comment by adamjstewart** [comment]
I would try installing the full XCode suite and launching XCode and signing that agreement. The command-line-tool-only package has given me trouble in the past.

**Comment by haampie** [comment]
Poor macOS users

**Comment by certik** [comment]
So when I install `make` using `spack` and ensure it is in PATH, then spack actually works fine in parallel.

Quite frankly, I don't think Spack should be using anything from the system. It's a good habit to assume it is just broken.

I thought maybe I have broken something on my previous laptop, so I didn't want to even bother reporting it. But when the exact same issue happens on a brand new M1 laptop, then I know this is an actual problem.

It can also be caused by some company specific setups, I am not going to go into details here. But spack is *designed* to work in exactly the environments that I am in. CC @tgamblin.

I will note that Conda just works --- I unpack, use the default install and everything just works, on both Intel and M1 Macs. The main reason I think is that it just does not depend on an kind of system stuff.

Solution: make Spack never to use system programs such as make. Except perhaps for bootstrapping. 

I have noticed that Spack unfortunately will use the system make, unless I provide the Spack build's make in PATH. I think that is a bug. Shouldn't Spack by default use it's own make?

**Comment by haampie** [comment]
> Quite frankly, I don't think Spack should be using anything from the system. It's a good habit to assume it is just broken.

I love this view. For linux I've created https://github.com/eth-cscs/spack-batteries-included which is based on the ideas of AppImage, basically a single-file executable that mounts a zstd-compressed squashfs file included in the binary as an immutable filesystem which contains not only spack, but also `python`, `make`, `git`, `ca-certificates-mozilla`, etc. If there's no fuse on your system, it still contains the tools to extract the squashfs file as if it were a tarball with `./spack.x --squashfs-extract` and allows you to call `./spack/bin/spack` directly.

However, it turns out, even if you go through all this trouble, you will find that build systems themselves include absolute paths to system executables, notably almost all autotools configure scripts pointing to /bin/file :( and file can't always deal with the file database provided by the squashfs file. Also, compilers are still required on the host system.

Also, there is: https://github.com/spack/spack/pull/25979

**Comment by certik** [comment]
The issue of bootstrapping a compiler from nothing is interesting, see here: https://github.com/certik/bcompiler/issues/1#issuecomment-793896926. It can be done and it should be done. But it is a lot of work. Note also that in terms of Spack, you not only need to bootstrap a compiler, but also all the basic utilities like tar, make (!), etc., all in the right order. In theory that can all be done.

So let's assume that will get done in the future.

Then the only remaining issues are indeed things like `/bin/file` and other such bugs to fix in each package.

**Comment by haampie** [comment]
I don't think a short-term goal is to have spack bootstrap ab initio :p currently we assume the os provides proper build tools, as well as a working python installation, which usually isn't too much to ask...

It's probably easier to bootstrap by downloading working binaries from somewhere (which we currently don't fully provide except for clingo), and then run `spack install spack`, like gcc builds gcc...

**Comment by certik** [comment]
(Yes, it is not a short term goal for me either. All I am saying is that it is interesting conceptually and eventually it would be nice to have.)

The clingo binary was not available on M1 (because it was trying to build it from source, and that's where it failed).

**Comment by haampie** [comment]
Even when you changed from urllib to curl?

In any case, the most likely outcome of pursuing bootstrapping from nothing is that Spack becomes a linux distro like NixOS, instead of just a package manager assuming a working distro, like Nix, which it is now.

**Comment by goxberry** [comment]
> Quite frankly, I don't think Spack should be using anything from the system. It's a good habit to assume it is just broken.

MacPorts uses this approach on macOS, last I checked. Depending on the use-case, I'm not sure there's economy in such an assumption because it trades off potential bugs in Spack-launched package builds for potential end-user configuration bugs when users try to use Spack-installed packages in their downstream software development. These latter bugs were what caused me to switch from using MacPorts to Homebrew in the first place; however, for HPC use cases involving bespoke MPI+BLAS+other software stacks, maybe the latter sort of configuration bugs are unavoidable.

Taken to the extreme, I'd also hate to saddle users with building large software packages from source, which seems difficult to avoid until binary buildcaches are available for more software packages. On more limited hardware, these build times can be substantial -- I vaguely remember building LLVM from source took around 6 hours on a 2014-era MacBook Air -- which makes using existing installations convenient when those installs aren't broken.

**Comment by haampie** [comment]
My only mac is a macbook air from 2015, I remember compiling julia from sources which requires building llvm took hours, before I had a proper desktop.

I still test spack every now and then on that macbook, python already feels terribly slow :p

**Comment by cjosey** [comment]
As a heads up, a similar "parallel macOS make doesn't work" was reported here: #21492.

So far, everyone who has reported similar issues in these threads works at the same place, which fuels my speculation that it is a misconfiguration on our end.

**Comment by certik** [comment]
I suspect that too that it is a misconfiguration on my end, although not my personal fault.

But at the end of the day, I've seen that over the past 20 years over and over that a system package is broken (at Linux or macOS), one way or another. And the solution is always to simply not depend on it. Whether in Sage, HashDist or now Spack. It's always the same story and always the same solution.

**Comment by goxberry** [comment]
> But at the end of the day, I've seen that over the past 20 years over and over that a system package is broken (at Linux or macOS), one way or another. And the solution is always to simply not depend on it. Whether in Sage, HashDist or now Spack. It's always the same story and always the same solution.

Bootstrapping aside, not depending on external software is *an* answer, but I don’t see how it could be *the* answer in all cases — e.g., at HPC sites, [it doesn’t make sense for users to build their own MPI](https://spack.readthedocs.io/en/latest/getting_started.html#mpi).

I could see the value in some sort of more expansive bootstrapping option in addition to the current paradigm of permitting externally-installed packages, if someone wants to implement that option and support it.

**Comment by certik** [comment]
Yes, you cannot use your own MPI at HPC sites, so that is an exception. You absolutely can and should build your own make, just like spack builds and uses its own cmake. Depending on system cmake would be similarly a bad idea, as typically it is too old, or can also be broken in various ways.

**Comment by alalazo** [comment]
CMake packages now depend on gmake explicitly since #35552 We are not completely independent of system `make` but we are going into that direction.

**Comment by alalazo** [comment]
Autotools packages depend on `gmake` too, see #40380 Closing this issue, but feel free to open a new one for specific packages.
